plugins/auth-backend/src/lib/OAuthProvider.ts

```diff
@@ -39,9 +39,41 @@ export type Options = {
   tokenIssuer: TokenIssuer;
 };
 
+/* Return the value of `env` key, if it is exists, encoded within
+     the `state` parameter in the request
+  */
+const readStateParameter = (
+  stateParams: Array<string>,
+  parameter: string,
+): string => {
+  const envParams = stateParams.filter(
+    param => param.split('=')[0] === parameter,
+  );
+
+  if (envParams.length > 0) {
+    return envParams[0].split('=')[1];
+  }
+  return '';
+};
+
+const readState = (stateString: string): Array<string> => {
+  return decodeURIComponent(stateString).split('&');
+};
+
+export const encodeState = (stateObject: object): string => {
+  const vals = Object.keys(stateObject).map(
+    key =>
+      `${encodeURIComponent(key)}=${encodeURIComponent(
+        Reflect.get(stateObject, key),
+      )}`,
+  );
+
+  return vals.join('&');
+};
 export const verifyNonce = (req: express.Request, providerId: string) => {
   const cookieNonce = req.cookies[`${providerId}-nonce`];
-  const stateNonce = req.query.state;
+  const state: Array<string> = readState(req.query.state?.toString() ?? '');
```

Comment:
Again, I feel this should just be a proper object type with named fields. Also is the toString necessary here?